File: src/duElementAPI/screen.lua
--[[    
    screen.lua 
    Copyright (C) 2020 Stephane Boivin (Devgeek studio enr.)
    
    This file is part of "DU offline sandbox API".

    "DU offline sandbox API" is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    "DU offline sandbox API" is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with "DU offline sandbox API".  If not, see <https://www.gnu.org/licenses/>.
]]
Iscreen = {id, name, screen, generic}

function Iscreen:new(o)
  o = o or {}
  setmetatable(o, self)
  self.__index = self

  o.maxHitPoints = 2000
  o.hitPoints = 2000
  o.mass = 134.064875
  o.class = 'ScreenUnit'
  o.integrity = 100
    
  o.screen = Screen:new{id = o.id, name = o.name}
  o.generic = genericElement:new{id = o.id, maxHitPoints = o.maxHitPoints, hitPoints = o.hitPoints, mass = o.mass, class = o.class, integrity = o.integrity }
 
  return o
end


function createInterfaceScreen(id, name)

  local obj = Iscreen:new{id=id,name=name}
  obj.activate = function() obj.screen:activate() end
  obj.addContent = function(px, py, html) obj.screen:addContent(px, py, html) end
  obj.clear = function() obj.screen:clear() end
  obj.activate = function() obj.screen:activate() end
  obj.deactivate = function() obj.screen:deactivate() end
  obj.getMouseX = function() return obj.screen:getMouseX() end  
  obj.getMouseY = function() return obj.screen:getMouseY() end  
  obj.getMouseState = function() return obj.screen:getMouseState() end
  
  obj.hide = function() obj.generic:hide() end
  obj.show = function() obj.generic:show() end
  obj.getData = function() obj.generic:getData() end
  obj.getId = function() return obj.generic:getId() end
  obj.getMass = function() return obj.generic:getMass() end
  obj.getElementClass = function() return obj.generic:getElementClass() end
  obj.getIntegrity = function() return obj.generic:getIntegrity() end
  obj.getHitPoints = function() return obj.generic:getHitPoints() end
  obj.getMaxHitPoints = function() return obj.generic:getMaxHitPoints() end
  obj._setMass = function(m) obj.generic:setMass(m) end
  obj._setMaxHitPoints = function(mhp) obj.generic:setMaxHitPoints(mhp) end
  obj._setHitPoints = function(hp) obj.generic:setHitPoints(hp) end
  obj._setIntegrity = function(int) obj.generic:setIntegrity(int) end
  
  return obj
end

Screen = {id, name = "Screen" }

function Screen:new(o)
  o = o or {}  
  setmetatable(o, self)
  self.__index = self
    
  if _verboseLUA == 1 then print("[LUA] Screen ["..o.name.."] created with the id ["..o.id.."]") end
    
  return o
end

function Screen:setSVG(svg)

end

function Screen:activate()
  if _verboseLUA == 1 then print("[LUA] Screen ["..self.name.."] is now active") end
  self:clear()
end

function Screen:deactivate()
  if _verboseLUA == 1 then print("[LUA] Screen ["..self.name.."] is now inactive") end
  self:clear()
end

function Screen:addContent(px, py, phtml) 
  self.html = phtml;
  -- print("[LUA] Screen ["..self.name.."] content changed")
  JavaWindow:setState(self.id, {"addContent", self.html, self.name})    
end

function Screen:getMouseX()
  return JavaWindow:getMouseX(self.id)
end

function Screen:getMouseY()
  return JavaWindow:getMouseY(self.id)
end

function Screen:getMouseState()
  return JavaWindow:getMouseState(self.id)
end


function Screen:clear()
  self.html = "";
  -- print("[LUA] Screen ["..self.name.."] cleared")
  JavaWindow:setState(self.id, {"clear", self.html, self.name})    
end

File: src/duJavaAPI/Screen.java
/*    
    Screen.java 
    Copyright (C) 2021 Stephane Boivin (Discord: Nmare418#6397)
    
    This file is part of "DU lua sandbox API".

    "DU lua sandbox API" is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    "DU lua sandbox API" is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with "DU lua sandbox API".  If not, see <https://www.gnu.org/licenses/>.
*/
package duJavaAPI;

import java.awt.Color;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.awt.event.MouseMotionAdapter;

import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.border.LineBorder;

import chrriis.dj.nativeswing.NSComponentOptions;
import chrriis.dj.nativeswing.swtimpl.components.JWebBrowser;
import sandbox.execWindow;


public class Screen extends BaseElement {
    public int sizeX;
    public int sizeY;
    public String html = "";
    private String cr = "\r\n";

//    String baseHtml = "<html><head></head><body></body></html>\r\n";
    
    String baseHtml = "<!DOCTYPE html>\r\n" + 
    		"<html>\r\n" + 
    		"<head>\r\n" + 
    		"<meta name=\"viewport\" content=\"width=device-width, initial-scale=1\">\r\n" + 
    		"<style>\r\n" + 
    		"#overlay {\r\n" + 
    		"  position: fixed;" + 
    		"  display: block;" + 
    		"  width: 100%%;" + 
    		"  height: 100%%;" + 
    		"  top: 0;" + 
    		"  left: 0;" + 
    		"  right: 0;" + 
    		"  bottom: 0;" + 
    		"  background-color: rgba(0,0,0,0);" + 
    		"  z-index: 2;" + 
    		"}\r\n" + 
    		"</style>\r\n</head><body><div style=\"color: #FFFFFF; background-color: #000000; width: 100%; height: 100%; \"> </div></body></html>\r\n";
    		
    // public Screen(int pid, String pname, int px, int py, String phtml) {this(pid, pname, px, py, 1014, 620, phtml);}    
	public Screen(int pid, String pname, int px, int py, int psizeX, int psizeY, boolean pverboseJava) {
		
		name = pname;
		x = px;
		y = py;
		sizeX = psizeX;
		sizeY = psizeY;
		id = pid;

		verboseJava = pverboseJava;
		
		// set the size
		// baseHtml = String.format(baseHtml, sizeX, sizeY);
		
		panel = new JPanel();		
		panel.setLayout(null);
		// panel.setVisible(false);
		panel.setBounds(px, py, psizeX+6, psizeY+25);
		panel.setBorder(LineBorder.createBlackLineBorder());
		panel.setBackground(Color.black);

		// lua require and interface 
		urlAPI = "src/duElementAPI/screen.lua";
        luaCall = pname+" = createInterfaceScreen(\""+pid+"\", \""+pname+"\")";
        
        if(verboseJava) System.out.println("[JAVA] Screen ["+name+"] created");	 		
	}

	    
	// Initialize a screen
	@Override
    public void LoadScreen(String phtml) {		
		 web = new JWebBrowser(JWebBrowser.destroyOnFinalization()); 
		// NSComponentOptions.constrainVisibility(),NSComponentOptions.destroyOnFinalization()
        web.setStatusBarVisible(false);
		web.setMenuBarVisible(false);
		// web.setBorder(null);		
		web.setBounds(1, 0, sizeX, sizeY);
		web.setButtonBarVisible(false);
		web.setBarsVisible(false);
		web.setAutoscrolls(false);
		web.setDoubleBuffered(true);
		web.setBackground(Color.black);
 
		
		web.setJavascriptEnabled(true);
		// web.navigate("https://www.dualthegame.com/fr/");
         
		web.getNativeComponent().addMouseMotionListener(new MouseMotionAdapter() {
			 @Override
			public void mouseMoved(MouseEvent e) { mouseX = e.getX(); mouseY = e.getY(); }
		});
		
		web.getNativeComponent().addMouseListener(new MouseAdapter() {
	        @Override
			public void mousePressed(MouseEvent me) { mouseState = 1; } /* System.out.println("Left Click!"); e.getX(), e.getY() */
	        @Override
			public void mouseReleased(MouseEvent me) { mouseState = 0; } /* System.out.println("released");  */
	    });
		web.setHTMLContent(baseHtml);
	
		panel.add(web, 1, 0);
		panel.setVisible(true);
		web.setHTMLContent(baseHtml);
    }

	@Override
    public void update(String[] param) {
//		String pname = param[2];
		String phtml = param[1];
		String pcommand = param[0];
		 
	    if(pcommand.equals("setSVG")) {

	    	// create an overlay
	    	html = "<div id=\"overlay\">"+phtml+"</div>";
	    	html = html.replace("\"", "\\\"");
	    	html = "document.getElementsByTagName(\"BODY\")[0].innerHTML = document.getElementsByTagName(\"BODY\")[0].innerHTML + \""+html+"\";";	    	
	    	html = html.replace( "\r", "");
	    	html = html.replace( "\n", "");	    	
	    	
	    	web.executeJavascript(html);
            // System.out.println("test:"+html);
			
	    }	
	    if(pcommand.equals("setCenteredText")) {
            
	    	String centerhtml = "<div style=\"color: #FFFFFF; font-size: 5VW; background-color: #000; width: 100%; height: 100%; text-align: center; line-height: 100VH;\">"+phtml+"</div>";	    
	   
	    	// create an overlay
	    	html = "<div id=\"overlay\">"+centerhtml+"</div>";
	    	html = html.replace("\"", "\\\"");
	    	html = "document.getElementsByTagName(\"BODY\")[0].innerHTML = document.getElementsByTagName(\"BODY\")[0].innerHTML + \""+html+"\";";	    	
	    	html = html.replace( "\r", "");
	    	html = html.replace( "\n", "");	    	
	    	
	    	web.executeJavascript(html);
            // System.out.println("test:"+html);
			
	    }	
	    if(pcommand.equals("addContent")) {

	    	// create an overlay
	    	html = "<div id=\"overlay\">"+phtml+"</div>";
	    	html = html.replace("\"", "\\\"");
	    	html = "document.getElementsByTagName(\"BODY\")[0].innerHTML = document.getElementsByTagName(\"BODY\")[0].innerHTML + \""+html+"\";";	    	
	    	html = html.replace( "\r", "");
	    	html = html.replace( "\n", "");	    	
	    	
	    	web.executeJavascript(html);
            // System.out.println("test:"+html);
	    }	
	    if(pcommand.equals("clear")) {
	    	web.executeJavascript("document.getElementsByTagName(\"BODY\")[0].innerHTML = \"\";");
	    }	
   		
	}


}
